walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -11,230 +11,188 @@
+    significantEndTimes.add(TimeRange.END_OF_DAY);
+    for (Event event: events) {
+      Integer startTime = event.getWhen().start();
+      Integer endTime = event.getWhen().end();
 
-      long durationRemaining = duration;
-      int minScore = 0;
-      int currTimeSlotIndex = startTimeSlotIndex;
-      
-      while ((durationRemaining > 0) && (currTimeSlotIndex < timeSegmentCount)) {
-        
-        Pair<TimeRange, Integer> timeRangeScorePair = timeRangeScorePairList.get(currTimeSlotIndex);
-        TimeRange timeRange = timeRangeScorePair.getValue0();
-        int timeSlotScore = timeRangeScorePair.getValue1();
-        
-        // If event duration spans more than 1 time slot
-        // We should record the lowest availabilty score of the span.
-        minScore = Math.min(minScore, timeSlotScore);  
+      significantEndTimes.add(startTime); // The commencement of a new period marks the end of another period.
+      significantEndTimes.add(endTime);
+    }
 
-        int timeSlotStartTime = timeRange.start();
-        int timeSlotEndTime = timeRange.end();
-        int timeSlotLength = timeSlotEndTime - timeSlotStartTime;
+    Integer begin = TimeRange.START_OF_DAY;
+    for (Integer end : significantEndTimes) {
+      if (end > begin) {
+        Boolean isInclusive = (end == TimeRange.END_OF_DAY) ? true : false;
+        timeRangeAttendanceList.add(TimeRangeAttendance.fromStartEnd(begin, end, isInclusive));
+      }
+      begin = end;
+    }
 
-        durationRemaining -= timeSlotLength;
+    Comparator < TimeRangeAttendance > c = TimeRangeAttendance.ORDER_BY_START;
 
-        currTimeSlotIndex++;
-      }
+    for (Event event: events) {
+      Integer startTime = event.getWhen().start();
+      Integer endTime = event.getWhen().end();
 
-      // We reached the end of the day
-      // but cannot find enough duration with a starting time of startTimeSlotIndex.
-      if (durationRemaining > 0) minScore = UNAVAILABLE_GUEST_PENALTY;
+      Collection < String > unavailableOptionalAttendees = getAttendeeOverlap(event.getAttendees(), optionalAttendees);
 
-      // Record the best score we have achieved.
-      bestAvailabilityScore = Math.max(bestAvailabilityScore, minScore);
-    }
+      Integer unavailableOptionalAttendeeCount = unavailableOptionalAttendees.size();
+      Integer unavailableMandatoryAttendeeCount = getAttendeeOverlap(event.getAttendees(), attendees).size();
 
-    return bestAvailabilityScore;
-  }
+      // Modify the the time ranges with penalties
+      if ((unavailableOptionalAttendeeCount > 0) || (unavailableMandatoryAttendeeCount > 0)) {
 
-  /* Return a list of TimeRange and associated Availability Score Pair from the previous TreeMap */
-  private ArrayList<Pair<TimeRange, Integer>> treeMapToTimeRangeScorePairList(TreeMap<Integer, Integer> timeCutoffs) {
+        Integer startIndex =
+          Collections.binarySearch(timeRangeAttendanceList, new TimeRangeAttendance(startTime, 0), c);
```

Comment:
This code is somewhat hard to follow. I think it would be cleaner if you looked for the starting index and then craft the for loop to end when timeRange.startTime >= event.timeEnd.

Overall I think the complexity will be similar to the approach I proposed, although you look at it from the other side and use binary search instead of sort.